### DISCUSSION TITLE
Making LRUCache capacity configurable

### DISCUSSION THREAD

**Discussion by vkomarov-r7** [sqlalchemy/sqlalchemy#7881] [Ideas]
In recent troubleshooting efforts, I've been looking at memory utilization of various SQLAlchemy constructs we have and encountered LRUCache as a complicating factor in multiple scenarios:

1. In our e2e test suite, we mistakenly called `create_engine` once for each test case. This didn't clear out existing LRUCache objects, causing us to falsely think we have a memory leak. This has been dealt with but was a huge cause of concern.
2. In a specific portion of our code, we resolve a lazy-loaded attribute from a separate table, this causes an LRU-cached Query object to be created for (presumably) all subsequent calls. Digging into it, I found the part where LazyLoader creates a Bakery instance, which points at an LRUCache of used queries.

It seems like the project uses caches as a first hand way to improve performance in a longer-running application, which has the drawback of looking like memory leaks in shorter-running applications (or test suites). As such, I was wondering if a few process improvements could be made:

1. Would it be possible to allow the end-user to configure the cache sizes of the various LRUCaches? Right now it seems like most (if not all) of them are hard-coded to a given value. When investigating memory leaks, this is the first thing I'd disable before looking further.
2. I noticed that LRU cache warnings were removed at some point in the past (https://github.com/sqlalchemy/sqlalchemy/commit/e83024d3fa45973d4dc1809282588dda8ac9bfce, with a very helpful commit message BTW!), would it make sense to conditionally re-enable them? I tracked the size of all of our LRU caches during a test run prior to the `create_engine` fix and found this, but this would be great to keep enabled in our lower environments as part of performance monitoring.

**Comment by vkomarov-r7** [comment]
For posterity, this is the `objgraph` that set it off for me for lazy loaded attributes.

![backrefs](https://user-images.githubusercontent.com/65631117/161154998-b44dff2e-d871-44ea-94cc-cd319b797c60.png)

**Reply by zzzeek** [reply]
I would recommend upgrading to 1.4.  Bakery objects are gone if you arent using them explicitly.     See https://docs.sqlalchemy.org/en/14/core/connections.html#sql-compilation-caching for a detailed overview of the new caching system.

**Reply by vkomarov-r7** [reply]
This looks amazing! Going to prioritize upgrading to 1.4 and see where we stand then. Thanks!

**Reply by zzzeek** [reply]
yeah, it should be better for sure.   you can at least turn it off and there's logging.  caching is extremely complicated nonetheless.

**Comment by zzzeek** [comment]
as mentioned in my comment, the Baked system is completely removed from any implicit use in the ORM and there is a Core/ORM-wide caching system documented at https://docs.sqlalchemy.org/en/14/core/connections.html#sql-compilation-caching .    While there was some churn with some details of how the ORM used the cache in early 1.4 versions, in later 1.4.x versions the cache that you pass for create_engine() is used exclusively for all SQL that's emitted.  if you set the cache size to zero, nothing caches including lazy loads.
